Client configuration

tg666 opened the pull request:
Hi,
when integrating the extension into our project, I found out that the Client configuration has an invalid structure and also that default (null) values are getting into the Client.

Current configuration example:

```neon
contributte.elastica:
	debug: %debugMode%
	hosts:
		es01:
			host: elasticsearch
			port: 9200
```

```php
<?php
​
public function createServiceContributte__elastica__client(): Contributte\Elastica\Client
{
	$service = new Contributte\Elastica\Client(
		[
			'es01' => [
				'host' => 'elasticsearch',
				'port' => 9200,
				'path' => null,
				'url' => null,
				'proxy' => null,
				'transport' => null,
				'persistent' => null,
				'timeout' => null,
				'connections' => [],
				'roundRobin' => null,
				'compression' => null,
				'log' => null,
				'retryOnConflict' => null,
				'bigintConversion' => null,
				'username' => null,
				'password' => null,
			],
		],
		null,
		$this->getService('contributte.monolog.logger.default')
	);
	$this->getService('contributte.elastica.panel')->register($service);
	return $service;
}
```

But options for the default Client should be in the configuration root. In the case of multiple nodes, they are defined under the key `connections`.

New configuration example:

```neon
contributte.elastica:
	debug: %debugMode%
	config:
		host: elasticsearch
		port: 9200
```

```php
public function createServiceContributte__elastica__client(): Contributte\Elastica\Client
{
	$service = new Contributte\Elastica\Client(
		['host' => 'elasticsearch`, 'port' => 9200],
		null,
		$this->getService('contributte.monolog.logger.default')
	);
	$this->getService('contributte.elastica.panel')->register($service);
	return $service;
}
```

I also marked options `host`, `port`, `username`, `password` and `auth_type` as dynamic - parameters, env variables etc. are allowed.

Package `nette/schema` is required because of usage of the method `::skipDefaults()` that has been added in the version `^1.2`.

tg666 committed:
Client configuration

- added dependency on package `nette/schema`
- fixed structure of config schema for the compiler extension `ElasticaExtension`
- removed an option `compression` (the option is not defined in `ClientConfiguration` class)
- removed an option `log` (the option has been removed in version `7.0`)
- added an option `auth_type`
- updated examples in README

f3l1x commented:
Hi. Does it means there is not support for multiple configuration?

tg666 commented:
Ruflin/Elastica supports multiple connections. They can be defined under key `connections`:
https://github.com/ruflin/Elastica/blob/master/src/Client.php#L619

For example:

```neon
contributte.elastica:
	debug: %debugMode%
	config:
		connections:
			-
				host: host1
				port: 1234
				# ...
			-
				host: host2
				port: 5678
				# ...
```

f3l1x commented on .docs/README.md:
```diff
@@ -29,36 +29,35 @@ Define at least one host, this would be minimal possible config.
 Take a look to [Elastica docs](https://elastica-docs.readthedocs.io/en/latest/client.html#client-configurations).
 
-## Usage
+## Usage**
```
Typo

tg666 replied:
Fixed

f3l1x commented on src/DI/ElasticaExtension.php:
```diff
@@ -20,29 +20,26 @@ class ElasticaExtension extends CompilerExtension
+				'transport' => Expect::string()->nullable(),
+				'persistent' => Expect::bool(),
+				'timeout' => Expect::int()->nullable(),
+				'connections' => Expect::array(), // host, port, path, transport, compression, persistent, timeout, username, password, auth_type, config -> (curl, headers, url)
```
Here must be array of structure with predefined keys. Can you put it here?

tg666 replied:
Done

f3l1x commented on src/DI/ElasticaExtension.php:
```diff
@@ -20,29 +20,26 @@ class ElasticaExtension extends CompilerExtension
+				'bigintConversion' => Expect::bool(),
+				'username' => Expect::string()->nullable()->dynamic(),
+				'password' => Expect::string()->nullable()->dynamic(),
+				'auth_type' => Expect::string()->nullable()->dynamic(), //basic, digest, gssnegotiate, ntlm
```
What about enum instead of string?

tg666 replied:
Good point, I changed it to an enum.

f3l1x commented:
Oh, I see it now. Thanks. I've added some comments.

tg666 committed:
Improved client configuration

- options for the default (root) client and also options under a key `connections` are strictly defined in a configuration schema
- updated the configuration keys in README

f3l1x commented on src/DI/ElasticaExtension.php:
```diff
@@ -20,29 +20,37 @@ class ElasticaExtension extends CompilerExtension
+			'config' => Expect::structure(array_merge(
+				$clientConfig,
+				[
+					'connections' => Expect::arrayOf(
```
Why is it merged here? Can we skip merging?

tg666 replied:
Because a client config contains 16 options and the options are present in the root `config` and also under the key `config.connections[]`. 
Merging came to be better for me. But yeah if you insist we can skip merging and define the options duplicated.

f3l1x commented on composer.json:
```diff
@@ -22,7 +22,8 @@
     "nette/di": "^3.0",
-    "nette/utils": "^3.0"
+    "nette/utils": "^3.0",
+    "nette/schema": "^1.2"
```
nette/schema is tight with nette/di, we can skip it here.

tg666 replied:
As I wrote in the first post:

> Package `nette/schema` is required because of usage of the method `::skipDefaults()` that has been added in the version `^1.2`.

All versions of `nette/di` requires `nette/schema` in versions `^1.0` or `^1.1` so here can be compatibility problem.
Another solution could be to remove `nette/schema` from requirements and add the section `conflict`:

```json
{
	"conflict": {
		"nette/schema": "<1.2"
	}
}
```

f3l1x commented:
Ping @tg666

jiripudil commented:
Hi, is there anything I can help with to get this merged? Correct me if I'm wrong (please do!) but I think the package is now unusable e.g. with docker-compose where ES is usually accessible via a different hostname than the default `localhost` 🤔

f3l1x commented:
It's OK, thanks for pinging me. @jiripudil
